### ISSUE TITLE
Address Form

### ISSUE THREAD

**Issue by m0ru** [researchstudio-sat/webofneeds#226]
**tl;dr** there's many different address formats, which is hard to handle in a UI.
# On Address Forms

The address form as it's sketched atm assumes an ecommerce-like address form with separate fields for street, city, country, postcode (or "postal code" / "zip code" / etc for some countries) as opposed to the alternative, a single field (like osm and google maps use). @cbusch assumes (an educated guess) that the former is easier to use for people with less tech experience (who're more likely to have used old-style addresses and -forms on letters and home-shopping than say gmaps). I couldn't find any empirical research on this and would love to try and compare the two alternatives in say an A/B test.

So the plan is to implement the classic-version (first). This will result in some technical difficulties which I'll try to sketch out and document below.

Here's a purportedly really good overview over different postal systems if you want to dive into it in depth: http://www.columbia.edu/~fdc/postal/
## Ux Patterns
- [Sean Rice on UxPatterns](http://uxpatterns.org/forms/address-forms/) recommends to put the zip code first (and 
- [UxMatters](http://www.uxmatters.com/mt/archives/2008/06/international-address-fields-in-web-forms.php)
- [ux.stackexchange-best-pattern-for-international-address-forms](http://ux.stackexchange.com/questions/6556/best-pattern-for-international-address-forms)
- [ux.stackexchange-combining-all-the-address-fields-into-one](http://ux.stackexchange.com/questions/22196/combining-all-the-address-fields-into-one)
- As a side-note - [Unified Forms](http://uxmovement.com/forms/why-users-fill-out-forms-faster-with-unified-text-fields/) -> we should lay out all fields in a straight column instead of next to each other.
## Issues to Tackle (Classical Form)
- Different internalization which don't necessarily run along language or even nation lines (e.g. zip code / post code / postal code / PLZ / 邮政编码)
- Different orderings: if want to support the like-analog-mail feeling as good as possible, we'd have to make the order of the fields also dependent on culture. [see here](http://www.uxmatters.com/mt/archives/2008/06/images/address_formats_fig4.png)
- Additional fields in other cultures, e.g. State & County (in the USA) which are required for uniqueness
- Different fields are optional or not (see above)
- Some fields that don't exist in other cultures. E.g. [Ireland doesn't have zips](http://www.iol.ie/~discover/mail.htm)  (they're only getting them this year), except for Dublin, which has district numbers. Many other countries still don't have them. South Korea doesn't use street names, only district + house number, etc...
- Nominatim doesn't support these individual formats to do address resolution (address -> lon, lat) but just assumes the U.S. system. Free-text search/resolution still works in all cases. However we'd have to try and parse/transform the results to fit our form to do auto-completion (which is a must-have imho)

**Comment by m0ru** [comment]
A rudimentary version of the single-line address-search-field has been implemented as of 11247fe, however the ux needs some polishing. I'd suggest it behaves as following: 

## Suggested User Workflow 

1) After the first word's been finished (a space-character has been entered) or when the user stops typing for 2 seconds the query get's send to the server
2) while the http-request is in progress the drop-down is visible and shows a spinning wheel (to communicate that it's working)
3) the request returns, the drop-down gets populated with matching addresses that nominatim could resolve
4a) continuing to type queries nominatim again (see (1)) 
4b) select one of the alternatives (using arrow-keys or the mouse)
5) the selected location replaces the query-text in the text-field and is shown on the map

## Only Resolvable Addresses (Design Question)

The system needs a location in the form of longitude and latitude. If the user doesn't select an entry from the drop-down those might not be defined. There are a two alternative ways we could deal with this:

* if none has been actively selected, use the first one. I'd put the blue highlight on the first entry in the drop-down to make that clear.
* not use the location when publishing a need unless it's a valid location
    * we could display a green checkmark or sthg similiar next to the search-field if a valid location has been selected (like the oebb page / quando does)

Personally i think the second would be cleaner and more predictable for the user (e.g. when deleting the text to reset the location input)

## Open TODOs

### Must-Haves

* Empty Result Sets: If the location can't be resolved (because it doesn't exist in any of the osm databases), we need to show some kind of error message, e.g. in the drop-down. 
* filter for duplicates 
* error handling (e.g. in case nominatim changes it's API)
    * make sure json-structure actually is as assumed
* fix the graphical glitches the markers' popups cause to the markers themselves
   * maybe changing the marker implementation helps.
   * might require patching the leaflet code

### Nice-to-Haves

* show all results on the map and highlight the one / zoom in on the one that the mouse hovers over (in the drop-down)
* highlight typed text in result strings (by placing <b>'s around it) 
* show spinner in folded out drop-down while http-request is loading 
* bind down button to open the dropdown (to prevent the browser's own autocompletion from showing/overlapping 
* hide results with a really low score behind a "show more"-fold? (to filter out bogus results)

**Comment by fkleedorfer** [comment]
We've iterated a lot on this
